libraries/AP_Torqeedo/AP_Torqeedo.h

```diff
@@ -12,358 +12,116 @@
+#include <GCS_MAVLink/GCS_MAVLink.h>
+
+#include "AP_Torqeedo_Params.h"
+
+// Maximum number of range finder instances available on this platform
+#define TORQEEDO_MAX_INSTANCES 3
+
+// first monitor is always the primary monitor
+#define TORQEEDO_PRIMARY_INSTANCE  0
+
+class AP_Torqeedo_Backend;
 
-#define TORQEEDO_MESSAGE_LEN_MAX    35  // messages are no more than 35 bytes
+class AP_Torqeedo 
+{
+    friend class AP_Torqeedo_Backend;
 
-class AP_Torqeedo : public AP_ESC_Telem_Backend {
 public:
     AP_Torqeedo();
 
+    // do not allow copies
     CLASS_NO_COPY(AP_Torqeedo);
 
-    static AP_Torqeedo* get_singleton();
+    // Torqeedo driver types
+    enum class Type {
+        None = 0,
 
-    // initialise driver
-    void init();
+#if AP_TORQEEDO_TQBUS_ENABLED
+        TQBus = 1,
+#endif
+    };
 
-    // consume incoming messages from motor, reply with latest motor speed
-    // runs in background thread
-    void thread_main();
+    enum class Status {
+        NotConnected = 0,
+        NoData = 1,
+        Good
+    };
 
-    // returns true if communicating with the motor
-    bool healthy();
+     // The TorqeedoState structure is filled in by the backend driver
+    struct Torqeedo_State { 
+        uint8_t instance;      // the instance number 
+        Status status;         // propellers status
+        const struct AP_Param::GroupInfo *var_info;
+    };
 
-    // run pre-arm check.  returns false on failure and fills in failure_msg
-    // any failure_msg returned will not include a prefix
-    bool pre_arm_checks(char *failure_msg, uint8_t failure_msg_len);
+    static const struct AP_Param::GroupInfo *backend_var_info[TORQEEDO_MAX_INSTANCES];
 
-    // report changes in error codes to user
-    void report_error_codes();
+    // parameters for each instance
+    static const struct AP_Param::GroupInfo var_info[];
 
-    // clear motor errors
-    void clear_motor_error() { _motor_clear_error = true; }
+    // detect and initialise any available propellers
+    void init();
 
-    // get latest battery status info.  returns true on success and populates arguments
-    bool get_batt_info(float &voltage, float &current_amps, float &temp_C, uint8_t &pct_remaining) const WARN_IF_UNUSED;
-    bool get_batt_capacity_Ah(uint16_t &amp_hours) const;
+    // update state of all propellers. Should be called at high rate from main loop
+    void update();
 
-    static const struct AP_Param::GroupInfo var_info[];
+    // return the number of propellers backends
+    uint8_t num_propellers() const { return num_instances; }
 
-private:
+    // return sensor type of a given instance
+    Type get_type(uint8_t instance) const;
 
-    // message addresses
-    enum class MsgAddress : uint8_t {
-        BUS_MASTER = 0x00,
-        REMOTE1 = 0x14,
-        DISPLAY = 0x20,
-        MOTOR = 0x30,
-        BATTERY = 0x80
-    };
+    // return sensor health
+    Status get_instance_status(uint8_t instance) const;
+    Status get_status() const;
```

Comment:
Let's move the status to the backend and change get_status(instance) to ask the backend.  This should also remove the need for the semaphore protecting the frontend state.